Conditionally lint macros and expansions

som-snytt opened the pull request:
Adds a flag `-Ywarn-macros`.

Don't lint macros unless they request it.

If warning, then traverse the original tree of
macro expansions to witness usages.

Fixes scala/bug#10296

som-snytt committed:
Conditionally lint macros and expansions

Adds a flag `-Ywarn-macros:none,before,after,both`.

Don't lint macros unless they request it.

If warning, then traverse the original tree of
macro expansions to witness usages.

Optionally do it again for expanded trees.

The default is to lint what the user wrote,
"before" expansion.

Fixes scala/bug#10296

xeno-by commented on src/compiler/scala/tools/nsc/settings/Warnings.scala:
```diff
@@ -17,6 +17,7 @@ trait Warnings {
 
   // Non-lint warnings.
 
+  val warnMacros           = BooleanSetting("-Ywarn-macros", "Enable warnings on macro expansions.")
```
Would it be possible to include both unexpanded and expanded code in lint checks?

som-snytt replied:
I didn't give a lot of thought to this. For this ticket, I think linting looks at both. The other use case is the ticket whether to emit warnings on unexpanded code. So maybe the space of options is none, both, unexpanded and expanded.

dragos commented:
Is there a downside in checking both original and expanded trees? Both usages should be counted alway, since the definition can't be removed otherwise. And if that's the case, there isn't a need for an additional flag either.

som-snytt commented on src/compiler/scala/tools/nsc/typechecker/TypeDiagnostics.scala:
```diff
@@ -492,14 +492,15 @@ trait TypeDiagnostics {
         )
 
-        override def traverse(t: Tree): Unit = {
+        override def traverse(t: Tree): Unit = if (settings.warnMacros || !hasMacroExpansionAttachment(t)) {
```
Does this mean: lint if `t` has nothing to do with a macro, or if they want linted macros?

som-snytt commented on src/compiler/scala/tools/nsc/typechecker/TypeDiagnostics.scala:
```diff
@@ -544,6 +545,8 @@ trait TypeDiagnostics {
           }
+          val expandee = macroExpandee(t)
+          if (expandee ne t) traverse(expandee)
           super.traverse(t)
```
Does this mean: if this is a macro and t is not the original expandee, then lint it; always lint the expanded. The flag is checked in traverse before actually descending into either tree. So it's either both trees or neither. Is that it?

lrytz commented:
@som-snytt what's the status here? There are some questions, also by yourself. Iulian suggests that the linter should check both trees (which it does as far when the new flag is enabled), and that we could just always do that, and not introduce the `-Ywarn-macros` flag. That sounds good to me, too. WDYT?

som-snytt commented:
I'll update this when I have a chance. My answer about multi choice, to lint both (by default), either or none of expandee and expanded, seems right to me. I'll try to motivate with a use case, but macro magic means a given macro might either consume or produce something that is spuriously warnable. Inevitably, someone will complain they can't use a macro lib because of linting; a flag is an escape hatch at least.

fommil commented:
I never realised how much I wanted this until I started turning linters / wartremover to the max :-)

som-snytt commented:
@adriaanm That was so easy, I thought I'd push my luck. After I take care of a small fire at the office, I'll re-review. I started looking yesterday with my new macro vision. But my home computer just died, etc.

som-snytt commented:
Pushed what I commented in July, default is to lint original trees, optionally lint again with expanded trees, or both, or neither.

The previous fix traversed both at once, which is less discriminating.

adriaanm committed:
Review #5876: OOPify

Good old overrides are sometimes easier to follow than
boolean flags and function arguments.

Also, incorporate som-snytt's review of the review.

adriaanm commented:
Turned my review into a commit. Looking GTY?

som-snytt commented:
One minor comment, otherwise thanks for another good idea.

I also didn't convert strings to the `MultiChoiceEnumeration` thing. Because lazy and pressed for time.

adriaanm commented:
The reviewer reviewed, as they say in French. Fixed the `traverse`, but turning the rest into TODOs.

som-snytt commented:
@adriaanm thanks, it's like The Flash showed up to help me finish vacuuming.

adriaanm commented:
Zoom zoom zoom! Or were you thinking of the black adder one?

stewSquared commented:
Can someone clarify what `both` means?
Suppose we have `-Ywarn-unused` and `-Ywarn-macros:both`
Does it error when a val is unused either before OR after macro expansion?
Or does it error only when a val is unused before AND after?

ie., does `both` generally result in more or fewer warnings?

som-snytt commented:
@stewSquared It runs the check both before and then after expansion, so you're liable to see more warnings.  The missing mode is "either", where a symbol could be used before but not after and conversely for other symbols. That would result in fewer warnings.

stewSquared commented:
I see. Thanks! I suppose `either` could be the most conservative option, but it look like it would be a bit trickier to implement.

This leaves `after` as the conservative option for most people. (`none` is then the least conservative, since spurious warnings can result from failure to find usages in either the expanded or unexpanded trees.)
